components/script/dom/indexeddb/idbtransaction.rs

```diff
@@ -176,27 +249,178 @@ impl IDBTransaction {
     /// associated `IDBOpenDBRequest` `"success"` event is fired afterwards.
     pub fn request_finished(&self) {
+        self.dec_pending_requests();
+        if self.aborted.get() {
+            self.maybe_finish_after_abort();
+        } else {
+            self.maybe_finish_successfully();
+        }
+    }
+
+    pub(crate) fn set_upgrade_metadata(&self, old_version: u64) {
+        self.upgrade_old_version.set(Some(old_version));
+    }
+
+    pub(crate) fn initiate_abort(&self, error: Error, can_gc: CanGc) {
+        // https://w3c.github.io/IndexedDB/#transaction-lifecycle
+        // A transaction can be aborted at any time before it is finished,
+        // even if the transaction isn’t currently active or hasn’t yet started.”
+        // Abort can be initiated any time before the transaction finishes.
+        // https://w3c.github.io/IndexedDB/#transaction-lifecycle
+        if self.finished.get() || self.aborted.get() {
+            return;
+        }
+
+        if self.error.get().is_none() {
+            if let Ok(exception) = create_dom_exception(&self.global(), error, can_gc) {
+                self.error.set(Some(&exception));
+            }
+        }
+
+        self.aborted.set(true);
+        let _ = self
+            .get_idb_thread()
+            .send(IndexedDBThreadMsg::AbortTransaction {
+                origin: self.global().origin().immutable().clone(),
+                name: self.db.get_name().to_string(),
+                transaction: self.serial_number,
+            });
+        // https://w3c.github.io/IndexedDB/#abort-a-transaction
+        // When a transaction is aborted the implementation must undo (roll back)
+        // any changes that were made to the database during that transaction.”
+        self.queue_abort_pending_requests();
+        // https://w3c.github.io/IndexedDB/#abort-a-transaction
+        if let Some(old_version) = self.upgrade_old_version.get() {
+            // Keep the DOM-side version consistent until backend cleanup runs.
+            self.db.set_version_cache(old_version);
+        }
+
+        // NOTE: Completion of abort waits for outstanding requests to finish.
         if self.pending_request_count.get() == 0 {
+            self.maybe_finish_after_abort();
+        }
+    }
+
+    fn queue_abort_pending_requests(&self) {
+        // https://w3c.github.io/IndexedDB/#transaction-concept
+        // A transaction has a request list of pending requests which have been made against the transaction.
+        // When a transaction is aborted the implementation must
+        // abort all outstanding requests made against the transaction.
+        let requests: Vec<Trusted<IDBRequest>> = self
+            .requests
+            .borrow()
+            .iter()
+            .map(|request| Trusted::new(&**request))
+            .collect();
+        let global = self.global();
+        global.task_manager().database_access_task_source().queue(
```

Comment:
Why queue a task since this is called already from a task? I'm guessing you need to first abort on the backend, and then send a message back to script to queue a task to fire the events.